README.md
<!--
# This file is automatically generated by a `metapak`
# module. Do not change it elsewhere, changes would
# be overridden.
-->
# cloud-pages
> Push your pages to cloud buckets (S3 for now).

[![NPM version](https://badge.fury.io/js/cloud-pages.svg)](https://npmjs.org/package/cloud-pages)
[![Build status](https://secure.travis-ci.org/nfroidure/cloud-pages.svg)](https://travis-ci.org/nfroidure/cloud-pages)
[![Dependency Status](https://david-dm.org/nfroidure/cloud-pages.svg)](https://david-dm.org/nfroidure/cloud-pages)
[![devDependency Status](https://david-dm.org/nfroidure/cloud-pages/dev-status.svg)](https://david-dm.org/nfroidure/cloud-pages#info=devDependencies)
[![Coverage Status](https://coveralls.io/repos/nfroidure/cloud-pages/badge.svg?branch=master)](https://coveralls.io/r/nfroidure/cloud-pages?branch=master)
[![Code Climate](https://codeclimate.com/github/nfroidure/cloud-pages.svg)](https://codeclimate.com/github/nfroidure/cloud-pages)
[![Dependency Status](https://dependencyci.com/github/nfroidure/cloud-pages/badge)](https://dependencyci.com/github/nfroidure/cloud-pages)

# API
## Functions

<dl>
<dt><a href="#initCloudPages">initCloudPages($, [name], [dependencies])</a> ⇒ <code>undefined</code></dt>
<dd><p>Declare <code>cloudPages</code> in the dependency injection system
and allow to change its name/dependencies</p>
</dd>
<dt><a href="#cloudPages">cloudPages(services, options)</a> ⇒ <code>Promise</code></dt>
<dd><p>Deploy pages to the cloud and optionnally remove old versions</p>
</dd>
</dl>

<a name="initCloudPages"></a>

## initCloudPages($, [name], [dependencies]) ⇒ <code>undefined</code>
Declare `cloudPages` in the dependency injection system
and allow to change its name/dependencies

**Kind**: global function  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| $ | <code>Knifecycle</code> |  | The knifecycle instance |
| [name] | <code>String</code> | <code>&#x27;cloudPages&#x27;</code> | The name of the service |
| [dependencies] | <code>Array</code> | <code>[&#x27;ENV&#x27;, &#x27;glob&#x27;, &#x27;fs&#x27;, &#x27;log&#x27;, &#x27;s3&#x27;, &#x27;mime&#x27;, &#x27;git&#x27;]</code> | The dependencies to inject |

<a name="cloudPages"></a>

## cloudPages(services, options) ⇒ <code>Promise</code>
Deploy pages to the cloud and optionnally remove old versions

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise to be resolved when the deployment ended  

| Param | Type | Description |
| --- | --- | --- |
| services | <code>Object</code> | Services (provided by the dependency injector) |
| services.ENV | <code>Object</code> | Environment service |
| services.glob | <code>function</code> | Globbing service |
| services.fs | <code>Object</code> | File system service |
| services.log | <code>function</code> | Logging service |
| services.s3 | <code>Object</code> | S3 bucket service |
| services.mime | <code>Object</code> | MIME mapping service |
| services.git | <code>Object</code> | Repository service |
| services.time | <code>function</code> | Time service |
| options | <code>Object</code> | Options (destructured) |
| options.last | <code>Number</code> | Number of prior versions to keep |
| options.delay | <code>Number</code> | Delay for keeping prior versions |
| options.dir | <code>String</code> | Directory that contains assets to deploy |
| options.gitDir | <code>String</code> | Directory of the git repository to look for versions |
| options.files | <code>String</code> | Pattern of files to deploy in the directory |
| options.ignore | <code>String</code> | Pattern of files to ignore in the directory |
| options.remove | <code>Boolean</code> | Boolean indicating if ld versions should be removed |
| options.version | <code>String</code> | Version for the current deployment |
| options.bucket | <code>String</code> | Targetted bucket |


# License
[MIT](https://github.com/nfroidure/cloud-pages/blob/master/LICENSE)


CHANGELOG.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/nfroidure/cloud-pages/compare/v1.0.0...v1.0.1) (2017-12-02)



<a name="1.0.0"></a>
# 1.0.0 (2017-04-12)


### Features

* **s3:** Basic Amazon S3 deployment ([41544c9](https://github.com/nfroidure/cloud-pages/commit/41544c9))





src/s3.js
'use strict';

const AWSS3 = require('aws-sdk/clients/s3');
const debug = require('debug')('s3');
const Promise = require('bluebird');

const DEFAULT_NAME = 's3';
const DEFAULT_DEPENDENCIES = ['ENV', 'log'];

module.exports = initS3Service;

function initS3Service(
  $,
  name = DEFAULT_NAME,
  dependencies = DEFAULT_DEPENDENCIES
) {
  $.service(name, $.depends(dependencies, s3Service));
}

function s3Service({ ENV, log }) {
  return Promise.resolve().then(() => {
    const s3 = new AWSS3({
      region: ENV.AWS_S3_REGION || ENV.AWS_REGION,
      accessKeyId: ENV.AWS_S3_ACCESS_KEY_ID || ENV.AWS_ACCESS_KEY_ID,
      secretAccessKey:
        ENV.AWS_S3_SECRET_ACCESS_KEY || ENV.AWS_SECRET_ACCESS_KEY,
      logger: { log: debug },
    });

    log('debug', 'S3 Service initialized.');

    return Promise.resolve({
      listObjectsAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.listObjectsV2(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
      deleteObjectsAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.deleteObjects(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
      createBucketAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.createBucket(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
      putObjectAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.putObject(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
      putBucketAclAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.putBucketAcl(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
      putBucketWebsiteAsync: (...args) =>
        new Promise((resolve, reject) => {
          s3.putBucketWebsite(...args, (err, data) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(data);
          });
        }),
    });
  });
}


src/git.js
'use strict';

const semver = require('semver');
const git = require('simple-git');
const Promise = require('bluebird');

const DEFAULT_NAME = 'git';
const DEFAULT_DEPENDENCIES = ['ENV', 'log'];

const TAGS_REGEXP = /\(([^)]+)\)$/;

module.exports = initGitService;

function initGitService(
  $,
  name = DEFAULT_NAME,
  dependencies = DEFAULT_DEPENDENCIES
) {
  $.service(name, $.depends(dependencies, gitService));
}

function gitService({ log }) {
  return Promise.resolve().then(() => {
    log('debug', 'Git Service initialized.');

    return Promise.resolve({
      getCommits: gitDir =>
        new Promise((resolve, reject) => {
          git(gitDir).log((err, result) => {
            if (err) {
              reject(err);
              return;
            }
            resolve(
              result.all
                .map(({ hash, date, message }) => {
                  const matches = TAGS_REGEXP.exec(message);
                  return {
                    hash,
                    time: new Date(date).getTime(),
                    message,
                    tags: matches
                      ? matches[1]
                          .split(',')
                          .map(s => s.trim())
                          .filter(s => s.startsWith('tag: '))
                          .map(s => s.substr('tag: '.length))
                          .filter(semver.valid)
                      : [],
                  };
                })
                .reduce(
                  (commits, { hash, time, message, tags }) =>
                    commits.concat(
                      tags.map(tag => ({ tag, hash, time, message }))
                    ),
                  []
                )
            );
          });
        }),
    });
  });
}


src/cloudpages.js
'use strict';

const YError = require('yerror');
const path = require('path');

/* Architecture Note #1.1: The `last` Option

The `last` option indicates how many versions should be
 kept in the cloud in order to smoothly change frontends
 avoiding users to face 404 errors if they still rely
 on an old frontend.
*/
const DEFAULT_LAST = 4;

/* Architecture Note #1.2: The `delay` Option

You can also specify a `delay` option indicating a delay
 for which versions should be kept anyway (in ms).

It default to a month and should be set to 0 to disable it.
*/
const DEFAULT_DELAY = 30 * 24 * 60 * 60 * 1000;

module.exports = initCloudPages;

/* Architecture Note #1: The `cloudPages` Service

The `cloudPages` service allows to deploy a new frontend version
 to the cloud by managing the number of prior versions to keep
 so that you won't have to manually check which version to keep
 and waste resources by forgoting to remove them.

This service needs some other services. To be able to mock and
 interchange them, we use
 [Knifecycle](https://github.com/nfroidure/knifecycle) for its
 dependency injection and inversion of control feature.

![Dependencies Graph](./DEPENDENCIES.mmd.png)
*/

/**
 * Declare `cloudPages` in the dependency injection system
 * and allow to change its name/dependencies
 * @param  {Knifecycle} $                  The knifecycle instance
 * @param  {String} [name='cloudPages']        The name of the service
 * @param  {Array}  [dependencies=['ENV', 'glob', 'fs', 'log', 's3', 'mime', 'git']]
 * The dependencies to inject
 * @returns {undefined}
 */
function initCloudPages(
  $,
  name = 'cloudPages',
  dependencies = ['ENV', 'glob', 'fs', 'log', 's3', 'mime', 'git', 'time']
) {
  $.service(
    name,
    $.depends(dependencies, services =>
      Promise.resolve(cloudPages.bind(null, services))
    )
  );
}

/**
 * Deploy pages to the cloud and optionnally remove old versions
 * @param {Object}   services        Services (provided by the dependency injector)
 * @param {Object}   services.ENV    Environment service
 * @param {Function} services.glob   Globbing service
 * @param {Object}   services.fs     File system service
 * @param {Function} services.log    Logging service
 * @param {Object}   services.s3     S3 bucket service
 * @param {Object}   services.mime   MIME mapping service
 * @param {Object}   services.git    Repository service
 * @param {Function} services.time   Time service
 * @param {Object}   options         Options (destructured)
 * @param {Number}   options.last    Number of prior versions to keep
 * @param {Number}   options.delay   Delay for keeping prior versions
 * @param {String}   options.dir     Directory that contains assets to deploy
 * @param {String}   options.gitDir  Directory of the git repository to look for versions
 * @param {String}   options.files   Pattern of files to deploy in the directory
 * @param {String}   options.ignore  Pattern of files to ignore in the directory
 * @param {Boolean}  options.remove  Boolean indicating if ld versions should be removed
 * @param {String}   options.version Version for the current deployment
 * @param {String}   options.bucket  Targetted bucket
 * @return {Promise}                 A promise to be resolved when the deployment ended
 */
function cloudPages(
  { ENV, glob, fs, log, s3, mime, git, time },
  {
    delay = DEFAULT_DELAY,
    last = DEFAULT_LAST,
    dir,
    gitDir,
    files = '**/*',
    ignore = '.git/**/*',
    version,
    remove = false,
    bucket,
  } = {}
) {
  return Promise.resolve()
    .then(() => {
      if ('string' !== typeof version) {
        throw new YError('E_VERSION_REQUIRED', typeof version, version);
      }
      if ('string' !== typeof dir) {
        throw new YError('E_DIR_REQUIRED', typeof dir, dir);
      }
      bucket = bucket || ENV.AWS_S3_BUCKET;
      if ('string' !== typeof bucket) {
        throw new YError('E_BUCKET_REQUIRED', typeof bucket, bucket);
      }
    })
    .then(() => {
      const pattern = path.join(dir, files);

      gitDir = gitDir || dir;

      log('info', 'Deploying ' + version);

      return glob(pattern, {
        dir,
        ignore: path.join(dir, ignore),
        dot: true,
        nodir: true,
        absolute: true,
      })
        .catch(err => {
          log('error', 'Directory scan failure:', pattern);
          log('stack', 'Stack:', err.stack);
          throw YError.wrap(err, 'E_SCAN_FAILURE', pattern);
        })
        .then(files => {
          log('debug', 'Succesfully scanned files', pattern);
          log('debug', 'Files:', files);
          if (0 === files.length) {
            throw new YError('E_NO_FILES', pattern);
          }
          return files;
        });
    })
    .then(files =>
      Promise.resolve(
        files.map(file => {
          const key = path.join(version, path.relative(dir, file));

          return Promise.resolve()
            .then(() => {
              log('debug', 'Sending file:', file, key);
              return s3
                .putObjectAsync({
                  Bucket: bucket,
                  ACL: 'public-read',
                  Key: key,
                  Body: fs.createReadStream(file),
                  ContentType: mime.lookup(file),
                })
                .then(() => {
                  log('debug', 'File sent:', file, key);
                });
            })
            .catch(err => {
              throw YError.wrap(err, 'E_S3_UPLOAD_ERROR', file, key);
            });
        })
      )
    )
    .then(() => {
      log('debug', 'Setting up the bucket ACLs.');
      return s3
        .putBucketAclAsync({
          Bucket: bucket,
          ACL: 'public-read',
        })
        .then(() => {
          log('debug', 'Bucket ACLs set.');
        });
    })
    .then(() => {
      log('debug', 'Setting up the website rules.');
      return s3
        .putBucketWebsiteAsync({
          Bucket: bucket,
          WebsiteConfiguration: {
            ErrorDocument: {
              Key: version + '/index.html',
            },
            IndexDocument: {
              Suffix: 'index.html',
            },
            // Always redirect to the current version
            RoutingRules: [
              {
                Redirect: {
                  ReplaceKeyWith: version + '/index.html',
                },
                Condition: {
                  KeyPrefixEquals: '/index.html',
                },
              },
            ],
          },
        })
        .then(() => {
          log('debug', 'Website rules set.');
        });
    })
    .then(() => {
      log('info', 'Successfully deployed ' + version);

      if (!remove) {
        return Promise.resolve();
      }

      log('info', 'Seeking for old versions to remove.');
      const now = time();
      return git.getCommits(gitDir).then(commits => {
        const versionsToRemove = commits.filter((commit, i) => {
          if (
            version !== commit.tag &&
            i + 1 > last &&
            commit.time < now - delay
          ) {
            return true;
          }
          return false;
        });

        log(
          'debug',
          'Found ' +
            versionsToRemove.length +
            ' versions to potentially remove.'
        );

        return Promise.all(
          versionsToRemove.map(
            _recursivelyDeleteVersionObjects.bind(
              null,
              { ENV, log, s3 },
              bucket
            )
          )
        );
      });
    });
}

function _recursivelyDeleteVersionObjects({ log, s3 }, bucket, commit) {
  log('debug', 'Listing objects for version "' + commit.tag + '"');
  return s3
    .listObjectsAsync({
      Bucket: bucket,
      Prefix: commit.tag + '/',
    })
    .then(objects => {
      if (!objects.Contents.length) {
        log('info', 'No objects found for version "' + commit.tag + '"');
        return Promise.resolve();
      }
      log('info', 'Removing objects for version "' + commit.tag + '"');
      log('debug', 'Files ', objects);
      return s3
        .deleteObjectsAsync({
          Bucket: bucket,
          Delete: {
            Objects: objects.Contents.map(({ Key }) => ({ Key })),
          },
        })
        .then(() => {
          if (objects.MaxKeys === objects.KeyCount) {
            return _recursivelyDeleteVersionObjects(
              { log, s3 },
              bucket,
              commit,
              true
            );
          }
          return Promise.resolve();
        });
    });
}


src/cloudpages.mocha.js
'use strict';

const assert = require('assert');
const sinon = require('sinon');
const Knifecycle = require('knifecycle').default;
const initCloudPages = require('./cloudpages');
const YError = require('yerror');

describe('cloudPages', () => {
  const TEST_BASE_TIME = new Date('2010-03-06T00:00:00Z').getTime();
  let $;

  beforeEach(() => {
    $ = new Knifecycle();
    $.constant('ENV', {});
    $.constant('glob', sinon.stub());
    $.constant('time', sinon.stub().returns(TEST_BASE_TIME));
    $.constant('fs', {
      createReadStream: sinon.stub(),
    });
    $.constant('mime', {
      lookup: sinon.stub(),
    });
    $.constant('git', {
      getCommits: sinon.stub(),
    });
    $.constant('s3', {
      putObjectAsync: sinon.stub(),
      putBucketWebsiteAsync: sinon.stub(),
      putBucketAclAsync: sinon.stub(),
    });
    $.constant('log', sinon.stub());
    initCloudPages($);
  });

  it('should init well', () =>
    $.run(['cloudPages']).then(({ cloudPages }) => {
      assert.equal(typeof cloudPages, 'function');
    }));

  it('should fail with no version', () =>
    $.run(['ENV', 'cloudPages']).then(({ ENV, cloudPages }) => {
      ENV.AWS_S3_BUCKET = 'lol';
      return cloudPages()
        .then(() => {
          throw new YError('E_NOT_SUPPOSED_TO_BE_HERE');
        })
        .catch(err => {
          assert.equal(err.code, 'E_VERSION_REQUIRED');
        });
    }));

  it('should fail with no dir', () =>
    $.run(['ENV', 'cloudPages']).then(({ ENV, cloudPages }) => {
      ENV.AWS_S3_BUCKET = 'lol';
      return cloudPages({
        version: '1.0.0',
      })
        .then(() => {
          throw new YError('E_NOT_SUPPOSED_TO_BE_HERE');
        })
        .catch(err => {
          assert.equal(err.code, 'E_DIR_REQUIRED');
        });
    }));

  it('should fail with no files found', () =>
    $.run(['ENV', 'cloudPages', 'glob']).then(({ ENV, cloudPages, glob }) => {
      ENV.AWS_S3_BUCKET = 'lol';
      glob.returns(Promise.resolve([]));
      return cloudPages({
        version: '1.0.0',
        dir: '../www',
      })
        .then(() => {
          throw new YError('E_NOT_SUPPOSED_TO_BE_HERE');
        })
        .catch(err => {
          assert.equal(err.code, 'E_NO_FILES');
        })
        .then(() => {
          assert.deepEqual(glob.args, [
            [
              '../www/**/*',
              {
                absolute: true,
                dir: '../www',
                dot: true,
                ignore: '../www/.git/**/*',
                nodir: true,
              },
            ],
          ]);
        });
    }));

  it('should work with a file found', () =>
    $.run(['ENV', 'cloudPages', 'glob', 'git', 's3', 'fs', 'mime']).then(
      ({ ENV, cloudPages, glob, git, s3, fs, mime }) => {
        const SIX_DAYS = 6 * 24 * 60 * 60 * 1000;

        ENV.AWS_S3_BUCKET = 'lol';
        glob.returns(Promise.resolve(['./index.html']));
        s3.putObjectAsync.returns(Promise.resolve());
        s3.putBucketAclAsync.returns(Promise.resolve());
        s3.putBucketWebsiteAsync.returns(Promise.resolve());
        fs.createReadStream.returns('read-stream');
        mime.lookup.returns('text/plain');
        git.getCommits.returns(
          Promise.resolve([
            {
              tag: 'v3.0.1',
              hash: '1fb2cd74a2f3957263d57de06962d19045769890',
              time: TEST_BASE_TIME - SIX_DAYS / 4,
              message: '3.0.0 (tag: v3.0.0)',
            },
            {
              tag: 'v3.0.0',
              hash: '1fb2cd74a2f3957263d57de06962d19045769890',
              time: TEST_BASE_TIME - SIX_DAYS / 3,
              message: '3.0.0 (tag: v3.0.0)',
            },
            {
              tag: 'v2.0.0',
              hash: '1fb2cd74a2f3957263d57de06962d19045769890',
              time: TEST_BASE_TIME - SIX_DAYS / 2,
              message: '3.0.0 (tag: v3.0.0)',
            },
            {
              tag: 'v2.0.0',
              hash: '1fb2cd74a2f3957263d57de06962d19045769890',
              time: TEST_BASE_TIME - SIX_DAYS * 2,
              message: '3.0.0 (tag: v3.0.0)',
            },
          ])
        );
        return cloudPages({
          keep: {
            last: 1,
            delay: SIX_DAYS,
          },
          dir: './dist',
          gitDir: '.',
          files: '**/*',
          ignore: 'node_modules/**/*',
          version: '1.0.0',
        }).then(() => {
          assert.deepEqual(glob.args, [
            [
              'dist/**/*',
              {
                absolute: true,
                dir: './dist',
                dot: true,
                ignore: 'dist/node_modules/**/*',
                nodir: true,
              },
            ],
          ]);
          assert.deepEqual(s3.putObjectAsync.args, [
            [
              {
                ACL: 'public-read',
                Body: 'read-stream',
                Bucket: 'lol',
                ContentType: 'text/plain',
                Key: 'index.html',
              },
            ],
          ]);
          assert.deepEqual(s3.putBucketAclAsync.args, [
            [
              {
                ACL: 'public-read',
                Bucket: 'lol',
              },
            ],
          ]);
          assert.deepEqual(s3.putBucketWebsiteAsync.args, [
            [
              {
                Bucket: 'lol',
                WebsiteConfiguration: {
                  ErrorDocument: {
                    Key: '1.0.0/index.html',
                  },
                  IndexDocument: {
                    Suffix: 'index.html',
                  },
                  RoutingRules: [
                    {
                      Condition: {
                        KeyPrefixEquals: '/index.html',
                      },
                      Redirect: {
                        ReplaceKeyWith: '1.0.0/index.html',
                      },
                    },
                  ],
                },
              },
            ],
          ]);
        });
      }
    ));
});
